autoparallel/graph_pp_runner.py

```diff
@@ -289,7 +420,45 @@ def stage_forward(
     # SEND/RECV op are avoided for special case with 2 adjacent stages on same rank
     # see [Note: V-schedule special case]
     if is_next_stage_on_this_rank:
-        stage_index_to_stage[stage_index + 1].set_local_fwd_input(output, mb_index)
+        stage_index_to_stage[stage.stage_index + 1].set_local_fwd_input(
+            output, mb_index
+        )
+
+
+def stage_forward(
+    action: _Action,
+    ctx: _PipelineContext,
+    numerics_logs: Optional[list[str]] = None,
+) -> None:
+    (
+        schedule,
+        stage_index_to_stage,
+        stage,
+        mb_index,
+        is_next_stage_on_this_rank,
+        is_prev_stage_on_this_rank,
+    ) = _prepare_fwd_common(action, ctx)
```

Comment:
i see why its this way now. but maybe you can 
(a) get any widely used objects off ctx into local vars inside stage_forward and pass those into the util functions
(b) just access the less widely used objects directly off ctx inside the utils to reduce the number of passed around locals / API surface